[CSS] Support @scope(:host)

mdubet opened the pull request:
#### dd222ded80fd90dcad76a793fde7357493ab2719
<pre>
[CSS] Support @scope(:host)
<a href="https://bugs.webkit.org/show_bug.cgi?id=294521">https://bugs.webkit.org/show_bug.cgi?id=294521</a>
<a href="https://rdar.apple.com/153443869">rdar://153443869</a>

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-shadow-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::isScopePseudoClass const):
* Source/WebCore/css/CSSSelector.h:

Helper function

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):

:scope can match :host

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Disable selector compilation for :scope

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchHostPseudoClassRules):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):

Allow traversing the Shadow DOM boundary and track it for matching

* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::hasHostOrScopePseudoClassSubjectInSelectorList):
(WebCore::Style::RuleSet::addRule):
(WebCore::Style::hasHostPseudoClassSubjectInSelectorList): Deleted.
* Source/WebCore/style/RuleSet.h:
(WebCore::Style::RuleSet::hasHostOrScopePseudoClassRulesInUniversalBucket const):
(WebCore::Style::RuleSet::hasHostPseudoClassRulesInUniversalBucket const): Deleted.

Add rules with :scope subject (which can potentially represent :host) to the special
case for :host subject

Canonical link: <a href="https://commits.webkit.org/297670@main">https://commits.webkit.org/297670@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/efe2e6cb97c1ceb23e42336a0e151b87b8d2140f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110728 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20823 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116755 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60996 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112691 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38977 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84196 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24818 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99714 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64637 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17849 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60550 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94211 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17908 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119545 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37768 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28067 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93164 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38153 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95979 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92988 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24054 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38020 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15766 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33747 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37664 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43136 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37326 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40665 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39034 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -680,8 +687,9 @@ std::pair<bool, std::optional<Vector<ElementRuleCollector::ScopingRootWithDistan
                             appendIfMatch(previousScopingRoot.get());
                     }
                 }
-                ancestor = ancestor->parentElement();
+
                 ++distance;
+                ancestor = ancestor->parentOrShadowHostElement();
```
Why the order swap?

mdubet replied:
Purely for prettiness. It looks cleaner to me to end the block with the condition update.

nt1m reviewed (approved):
I don't think it conflicts with the `:scope` resolution here but worth checking: https://github.com/w3c/csswg-drafts/issues/11000#issuecomment-2622719814

mdubet committed:
[CSS] Support @scope(:host)
https://bugs.webkit.org/show_bug.cgi?id=294521
rdar://153443869

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-shadow-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::isScopePseudoClass const):
* Source/WebCore/css/CSSSelector.h:

Helper function

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):

:scope can match :host

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Disable selector compilation for :scope

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchHostPseudoClassRules):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):

Allow traversing the Shadow DOM boundary and track it for matching

* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::hasHostOrScopePseudoClassSubjectInSelectorList):
(WebCore::Style::RuleSet::addRule):
(WebCore::Style::hasHostPseudoClassSubjectInSelectorList): Deleted.
* Source/WebCore/style/RuleSet.h:
(WebCore::Style::RuleSet::hasHostOrScopePseudoClassRulesInUniversalBucket const):
(WebCore::Style::RuleSet::hasHostPseudoClassRulesInUniversalBucket const): Deleted.

Add rules with :scope subject (which can potentially represent :host) to the special
case for :host subject

Canonical link: https://commits.webkit.org/297670@main
